cmd/grpc-client-cli/msgbuffer.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+	"github.com/vadimi/grpc-client-cli/internal/caller"
+	"gopkg.in/AlecAivazis/survey.v1/terminal"
+)
+
+type msgBuffer struct {
+	opts       *msgBufferOptions
+	fieldNames []string
+	// next message prompt
+	nextPrompt string
+	helpText   string
+}
+
+type msgBufferOptions struct {
+	reader      *msgReader
+	messageDesc *desc.MessageDescriptor
+}
+
+func newMsgBuffer(opts *msgBufferOptions) *msgBuffer {
+	return &msgBuffer{
+		nextPrompt: "Next message (press Ctrl-D to finish): ",
+		opts:       opts,
+		fieldNames: fieldNames(opts.messageDesc),
+		helpText:   getMessageDefaults(opts.messageDesc),
+	}
+}
+
+func (b *msgBuffer) ReadMessage(opts ...ReadLineOpt) ([]byte, error) {
+	for {
+		message, err := b.opts.reader.ReadLine(b.fieldNames, opts...)
+		if err != nil {
+			if err == terminal.InterruptErr {
+				return nil, terminal.InterruptErr
+			}
+			return message, err
+		}
+
+		normMsg := bytes.TrimSpace(message)
+		if len(normMsg) > 0 {
+			if bytes.Equal(normMsg, []byte("?")) {
+				fmt.Println(b.helpText)
+				continue
+			}
+
+			if err := b.validate(normMsg); err != nil {
+				fmt.Println(err)
+				continue
+			}
+			return normMsg, nil
+		}
+	}
+}
+
+func (b *msgBuffer) ReadMessages() ([][]byte, error) {
+	if b.opts == nil || b.opts.reader == nil {
+		return nil, errors.New("no msg reader is configured")
+	}
+
+	msg, err := b.ReadMessage()
+	if err != nil {
+		return nil, err
+	}
+
+	buf := [][]byte{msg}
+
+	for {
+		msg, err := b.ReadMessage(WithReadLinePrompt(b.nextPrompt))
+		if err == terminal.InterruptErr {
+			return nil, terminal.InterruptErr
+		}
+
+		// Ctrl+D will trigger io.EOF if the line is empty
+		// it means no new messages are expected
+		if err == io.EOF {
+			return buf, nil
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		buf = append(buf, msg)
```

Comment:
there is a reader buffer size in msgreader.go 

Os there any size constraint we need to consider in the buffer ?

Reply:
this buffer is used to buffer messages for input stream (client or bi-directional streaming), I don't expect anyone would manually add so many messages that the app would run out of memory